pcdsdevices/tpr.py

```diff
@@ -0,0 +1,138 @@
+from enum import Enum
+
+from ophyd import Component as Cpt
+from ophyd import Device, EpicsSignal, EpicsSignalRO
+from ophyd import FormattedComponent as FCpt
+
+from .interface import BaseInterface
+from .pv_positioner import PVPositionerIsClose
+from .signal import MultiDerivedSignal, MultiDerivedSignalRO
+from .type_hints import SignalToValue
+from .variety import set_metadata
+
+TPR_TICK_NS = 5.384
+TPR_TAP_NS = 0.08
+
+
+class TimingMode(Enum):
+    LCLS1 = 1
+    LCLS2 = 2
+
+
+def _get_delay(mds: MultiDerivedSignal, items: SignalToValue) -> float:
+    """Calculates delay in ns from ticks and taps"""
+    return items[mds.attrs[0]] * TPR_TICK_NS + items[mds.attrs[1] * TPR_TAP_NS]
+
+
+def _get_width(mds: MultiDerivedSignal, items: SignalToValue) -> float:
+    """Calculates width in ns from ticks"""
+    return items[mds.attrs[0]] * TPR_TICK_NS
+
+
+def _put_last(mds: MultiDerivedSignal, value: float) -> SignalToValue:
+    """Only writes value to last attr"""
+    return {mds.attrs[-1]: value}
+
+
+class TprMotor(PVPositionerIsClose):
+    """
+    PV Positioner for adjusting an TPR channel.
+
+    Moves that are less than one tick
+    are considered immediately complete.
+    """
+    setpoint = FCpt(EpicsSignal, "{self.prefix}{self.sys}TDES", kind="normal", doc="Trigger delay setpoint in nsec")
+    delay_ticks = FCpt(EpicsSignal, '{self.prefix}TDESTICKS', kind="omitted", doc="Trigger delay in clock ticks")
+    delay_taps = FCpt(EpicsSignal, '{self.prefix}TDESTAPS', kind="omitted", doc="Trigger delay in delay taps")
```

Comment:
You don't need to change your code for this, but you should know that:

- For `FormattedComponent`, you don't need to include `self.` in the attributes. This used to be required but no longer is.
- If a component only needs the prefix, it can simply be a normal component, e.g. this last line is equivalent to `Cpt(EpicsSignal, 'TDESTAPS', ...)`

I think using the more verbose forms intentionally can be reasonable for clarity, but you should know the options.

Reply:
Good to know.